bibliopixel/led/coord_map.py

```diff
@@ -0,0 +1,143 @@
+
+    result = []
+    for y in range(dy):
+        if not serpentine or y % 2 == 0:
+            result.append([(dx * y) + x + offset for x in range(dx)])
+        else:
+            result.append([((dx * (y + 1)) - 1) - x + offset for x in range(dx)])
+
+    result = rotate_and_flip(result, rotation, y_flip)
+
+    return result
+
+
+def gen_cube(dx, dy, dz, xy_serpentine=True, offset=0,
+             xy_rotation=Rotation.ROTATE_0, z_rotation=Rotation.ROTATE_0,
+             y_flip=False, z_flip=False):
+    result = []
+    plane_offset = offset
+    for z in range(dz):
+        plane = gen_matrix(dx, dy, serpentine=xy_serpentine,
+                           offset=plane_offset, rotation=xy_rotation,
+                           y_flip=y_flip)
+        plane_offset += (dx * dy)
+        result.append(plane)
+
+    result = rotate_and_flip(result, z_rotation, z_flip)
+
+    return result
+
+
+def point_list_from_cube(coord_map):
+    num = 0
+    dx, dy, dz = (0, 0, len(coord_map))
+
+    for xy in coord_map:
+        if len(xy) > dx:
+            dy = len(xy)
+        for x in xy:
+            if len(x) > dx:
+                dx = len(x)
+            num += len(x)
+
+    result = [None] * num
+    for z in range(dz):
+        for y in range(dy):
+            for x in range(dx):
+                result[coord_map[z][y][x]] = [x, y, z]
+    return result
+
+
+def calc_ring_pixel_count(rings):
+    num = 0
+    for r in rings:
+        num += len(r)
+    return num
+
+
+def calc_ring_steps(rings):
+    steps = []
+    for r in rings:
+        steps.append(360.0 / len(r))
+    return steps
+
+
+def gen_circle(rings=None, pixels_per=None, offset=0, invert=False):
+    if pixels_per:
+        rings = []
+        for c in pixels_per:
+            rings.append([offset, offset + c])
+            offset = offset + c
+
+    if not rings:
+        raise ValueError('Must specify rings or pixels_per')
+
+    if rings:
+        num = 0
+        out_rings = []
+        for r in rings:
+            if len(r) != 2:
+                raise ValueError('"rings" values must only be first and last index.')
+            if r[0] < r[1]:
+                indices = list(range(r[0], r[1]))
+            else:
+                indices = list(range(r[1], r[0]))[::-1]
+            out_rings.append(indices)
+            num += len(indices)
+
+        if invert:
+            out_rings = out_rings[::-1]
+
+        return (out_rings, calc_ring_steps(out_rings))
+
+
+def point_list_from_rings(rings, origin=(0, 0, 0), z_diff=0):
+    if len(origin) not in [2, 3]:
+        raise ValueError('origin must be (x,y) or (x,y,z)')
+
+    use_z = len(origin) == 3
+
+    num = calc_ring_pixel_count(rings)
+    steps = calc_ring_steps(rings)
+
+    points = [None] * num
+    z = 0
+    for i in range(len(rings)):
+        r = rings[i]
+        step = steps[i]
+        angle = 0.0
+        for p in r:
+            radius = (len(r) * 3)

```

Comment:
lol... that was a shitty attempt to get it close without having floating point coordinates as SimPixel doesn't handle those. Only int.